scipy/linalg/decomp_svd.py

```diff
@@ -238,3 +238,101 @@ def orth(A):
     num = numpy.sum(s > tol, dtype=int)
     Q = u[:, :num]
     return Q
+
+
+def subspace_angles(A, B):
+    r"""
+    Compute the subspace angles between two matrices.
+
+    Parameters
+    ----------
+    A : (M, N) array_like
+        The first input array.
+    B : (M, K) array_like
+        The second input array.
+
+    Returns
+    -------
+    angles : ndarray, shape (min(N, K),)
+        The subspace angles between the column spaces of `A` and `B`.
+
+    See Also
+    --------
+    orth
+    svd
+
+    Notes
+    -----
+    This computes the subspace angles according to the formula
+    provided in [1]_. For equivalence with MATLAB and Octave behavior,
+    use ``angles[0]``.
+
+    .. versionadded:: 1.0
+
+    References
+    ----------
+    .. [1] Knyazev A, Argentati M (2002) Principal Angles between Subspaces
+           in an A-Based Scalar Product: Algorithms and Perturbation
+           Estimates. SIAM J. Sci. Comput. 23:2008-2040.
+
+    Examples
+    --------
+    A Hadamard matrix, which has orthogonal columns, so we expect that
+    the suspace angle to be :math:`\frac{\pi}{2}`:
+
+    >>> from scipy.linalg import hadamard, subspace_angles
+    >>> H = hadamard(4)
+    >>> print(H)
+    [[ 1  1  1  1]
+     [ 1 -1  1 -1]
+     [ 1  1 -1 -1]
+     [ 1 -1 -1  1]]
+    >>> np.rad2deg(subspace_angles(H[:, :2], H[:, 2:]))
+    array([ 90.,  90.])
+
+    And the subspace angle of a matrix to itself should be zero:
+
+    >>> subspace_angles(H[:, :2], H[:, :2]) <= 2 * np.finfo(float).eps
+    array([ True,  True], dtype=bool)
+
+    The angles between non-orthogonal subspaces are in between these extremes:
+
+    >>> x = np.random.RandomState(0).randn(4, 3)
+    >>> np.rad2deg(subspace_angles(x[:, :2], x[:, [2]]))
+    array([ 55.832])
+    """
+    A = _asarray_validated(A, check_finite=True)
+    if len(A.shape) != 2:
+        raise ValueError('expected matrix')
+    B = _asarray_validated(B, check_finite=True)
+    if len(B.shape) != 2:
+        raise ValueError('expected matrix')
+    if len(B) != len(A):
+        raise ValueError('A and B must have the same number of rows')
+
```

Comment:
I think a more informative check can be given here to help the user what the problem is. I've also noticed that when the error mentions a matrix, the users tend to confuse it with `np.matrix`. I can propose a check along the lines below. Since we also don't know what to do with `ndim>2` it might be a better idea to constrain it to exactly to 2D arrays.

    for ind, x in enumerate((A,B)):
        if not x.ndim == 2:
            raise ValueError('Array {} must be a 2D array'.format('AB'[ind]))
    if B.shape[0] != A.shape[0]:
        raise ValueError('A and B must have the same number of rows')